ClickHouse/ClickHouse
# Support disrtributed hive cluster query

Changelog category (leave one):
- New Feature



Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add storage engine `HiveCluster` and table funtion `hiveCluster` to support distribted query for hive.

query by table funtction `hiveCluster`
```sql
SELECT 
    day, count(*) 
FROM hiveCluster(
    'cluster_simple', -- clickhouse cluster
    'thrift://hivetest:9083', -- hive metastore url
    'tmp', -- hive table's database
    'demo', -- hive table's name
    'id Nullable(String), score Nullable(Int32), day String',  -- table structure
    'day' -- partition keys, comma separated
) 
GROUP BY day ORDER BY day
```

query by storage `HiveCluster`
```sql
CREATE TABLE IF NOT EXISTS default.demo_parquet ON CLUSTER cluster_simple (
    `id` Nullable(String), `score` Nullable(Int32), `day` String -- table structure
) 
ENGINE = HiveCluster(
    'cluster_simple', -- clickhouse cluster
    'thrift://hivetest:9083', -- hive metastore url
    'test', -- hive table's databse
    'demo' -- hive table's name
)
PARTITION BY(day);

SELECT day, count(*) FROM default.demo_parquet GROUP BY day ORDER BY day
```

Some big changes to notice
1. `StorageHiveCluster` is based on `StorageHive`, and `StorageHive` has a little big change. Move the hive files collect action in `StorageHive` into a single class `LocalHiveSourceFilesCollector`


Some benchmark results
1. run group by query by using storage engine `Hive` on single node
```
11 rows in set. Elapsed: 7.728 sec. Processed 119.83 million rows, 24.63 GB (15.51 million rows/s., 3.19 GB/s.

```
2. run group by query by using storage engine `HiveCluster` on three nodes cluster
```
11 rows in set. Elapsed: 3.745 sec. Processed 119.83 million rows, 24.63 GB (32.00 million rows/s., 6.58 GB/s.)
```


> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/
